### ISSUE TITLE
Python 3.5 support seems broken

### ISSUE THREAD

**Issue by LogicalDash** [jruizgit/rules#45]
I got durable to install in a Python 3.5 virtualenv, but trying to `import durable.lang` first gives me a couple of import errors, eg.`ImportError: No module named 'engine'` which are easily fixed by adding a `from .` on the start of the line, but then when durable tries to `import rules` I get

```
ImportError: /home/sanotehu/durable-rules/lib/python3.5/site-packages/rules.cpython-35m-x86_64-linux-gnu.so: undefined symbol: Py_InitModule
```

That seems to point to something wrong in the C module.

**Comment by jruizgit** [comment]
Hi, thanks a lot for reporting this issue. durable_rules doesn't support Python 3.x yet.

**Comment by aaronspringer-parc** [comment]
Any idea how much work it would take to get Python 3 support up and running?

The error referenced is pretty trivial to fix by replacing initrules in rules.c with
```
static struct PyModuleDef cModPyDem =
{
    PyModuleDef_HEAD_INIT,
    "durable_rules",
    "",
    -1,
    myModule_methods
};

PyMODINIT_FUNC PyInit_rules(void)
{
    return PyModule_Create(&cModPyDem);
}
```
This enables simple programs to work with Python3 but eventually runs into issues with operations between avalues and Python number types.

If you have any insight into how much work this transition would be that would be greatly appreciated. Thanks!

**Comment by jruizgit** [comment]
Hi, thanks for posting the question. It should take 3-5 days to get python 3.5 up and running and tested. It might take a couple more to get it up and running in Windows. Let me prioritize this work as it will unblock interesting use cases.


On Jan 24, 2017, at 2:50 PM, aaronspringer-parc <notifications@github.com<mailto:notifications@github.com>> wrote:


Any idea how much work it would take to get Python 3 support up and running?

The error referenced is pretty trivial to fix by replacing initrules in rules.c with

static struct PyModuleDef cModPyDem =
{
    PyModuleDef_HEAD_INIT,
    "durable_rules",
    "",
    -1,
    myModule_methods
};

PyMODINIT_FUNC PyInit_rules(void)
{
    return PyModule_Create(&cModPyDem);
}


This enables simple programs to work with Python3 but eventually runs into issues with operations between avalues and Python number types.

If you have any insight into how much work this transition would be that would be greatly appreciated. Thanks!

-

**Comment by jruizgit** [comment]
Hi, reporting back on this. I have implemented python 3.5/3.6 support. It is now available in pip for Mac and Linux. I'm still working on Windows support.

Sent from my iPhone

On Jan 24, 2017, at 2:50 PM, aaronspringer-parc <notifications@github.com<mailto:notifications@github.com>> wrote:


Any idea how much work it would take to get Python 3 support up and running?

The error referenced is pretty trivial to fix by replacing initrules in rules.c with

static struct PyModuleDef cModPyDem =
{
    PyModuleDef_HEAD_INIT,
    "durable_rules",
    "",
    -1,
    myModule_methods
};

PyMODINIT_FUNC PyInit_rules(void)
{
    return PyModule_Create(&cModPyDem);
}


This enables simple programs to work with Python3 but eventually runs into issues with operations between avalues and Python number types.

If you have any insight into how much work this transition would be that would be greatly appreciated. Thanks!

-

**Comment by aaronspringer-parc** [comment]
Wow, that's amazing! Thank you so much!

I was asking because I had 3.5 running but was having issues with one of the test cases and was curious if there were some more pernicious things I was missing. This is an awesome surprise to come back to from the weekend!
